core/src/main/java/io/grpc/transport/AbstractStream.java

```diff
@@ -229,6 +270,30 @@ protected final void requestMessagesFromDeframer(int numMessages) {
     }
   }
 
+  /**
+   * Event handler to be called by the subclass when a number of bytes are being queued for sending
+   * to the remote endpoint.
+   *
+   * @param numBytes the number of bytes being sent.
+   */
+  protected final void onSendingBytes(int numBytes) {
+    numSentBytesQueued += numBytes;
+  }
+
+  /**
+   * Event handler to be called by the subclass when a number of bytes has been sent to the remote
+   * endpoint. This method will invoke {@link StreamListener#onReady(int)} when appropriate.
+   *
+   * @param numBytes the number of bytes that were sent.
+   */
+  protected final void onSentBytes(int numBytes) {
+    numSentBytesQueued -= numBytes;
+    if (numSentBytesQueued < onReadyThreshold && !isClosed()) {
+      // TODO(nathanmittler): Consider selecting a better value for numMessages.
+      listener().onReady(1);
```

Comment:
I tried to communicate this earlier, but this doesn't work when an application exceeds the number of tokens provided.

1. We call `onReady(1)` (token total: 1)
2. Application sends two messages, each greater than `onReadyThreshold` (token total: -1)
3. After sending, eventually we call `onReady(1)` (token total: 0)
4. Deadlock, as application is waiting for another token.

There are two solutions:

1. Get rid of tokens. Just have zero-arg `onReady()`. If the application hasn't sent after the last onReady(), they can send more. If they sent more than once consecutively, then the application will over-buffer, but everything continues functioning.
2. Count how many messages are in each frame, and make sure to return that number of tokens.

Netty's outbound flow control is like option 1 (and so that is what my "streams as Channels" PR would provide... eventually).